### PR TITLE
Improve pom detection algorithm

### PR DESCRIPTION
**Pull request by arangatang**
## Types of changes
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Description
Using `VfsUtil.collectChildrenRecursively` can be prohibitively slow for very large projects making the validation take a very long time. This pr contains these changes to speed this up:

1. Uses kotlins walkTopDown which allows for early stopping of recursion when `pom.xml` is found.
2. Limits the max depth we search for `pom.xml` to some limit (currently 5 dirs deep, previously unlimited)
3. Avoids running the search for `pom.xml` twice as was done before. 

These changes makes a ~200 MB with depth of 12 take <1 second to scan instead of ~10 before.

### Before

https://github.com/aws/aws-toolkit-jetbrains/assets/9272897/da6bb10c-660e-4d70-989b-db6bae91c6ad

### After

https://github.com/aws/aws-toolkit-jetbrains/assets/9272897/7b86eb12-aff3-4e22-a659-46b1f4b21cca

## Checklist
- [X] My code follows the code style of this project
- [X] I have added tests to cover my changes
- [X] A short description of the change has been added to the **[CHANGELOG](https://github.com/aws/aws-toolkit-jetbrains/blob/master/CONTRIBUTING.md#contributing-via-pull-requests)** if the change is customer-facing in the IDE.
- [ ] I have added metrics for my changes (if required)
 
## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by dhasani23 — COMMENTED** [review]
Just double check if we want to search for pom.xml in root dir only

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt [hunk 1]
```diff
@@ -638,6 +639,29 @@ class CodeModernizerManager(private val project: Project) : PersistentStateCompo
...
     }
 
+    private fun findBuildFiles(sourceFolder: File): List<File> {
+        /**
+         * For every dir, check if any supported build files (pom.xml etc) exists.
+         * If found store it and stop further recursion.
+         */
+        val buildFiles = mutableListOf<File>()
+        sourceFolder.walkTopDown()
+            .maxDepth(5)
```
Comment (rli) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt:649]: is this a good value for the depth parameter?

**Review by rli — APPROVED** [review]

**Comment by arangatang** [comment]
@dhasani23 
We should be searching deeper than the root only, but the uploaded zip should have a pom in the root directory.
Further if the poms are nested then the top level pom should be the only option for customers.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt (hunk 1 above)
Comment (damntrecky) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt:649]: I'm not a good source for Java/Kotlin so I don't know what is a reasonable depth for these types of projects. Where would be a good source for this?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt [hunk 2]
```diff
@@ -638,6 +639,29 @@ class CodeModernizerManager(private val project: Project) : PersistentStateCompo
         managerState.flags.putAll(state.flags)
     }
 
+    private fun findBuildFiles(sourceFolder: File): List<File> {
```
Comment (damntrecky) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt:642]: Question: Do we have a test to make sure this doesn't find the file after N depth?

Would this be too much effort to create and not a lot of value proposition in that test?

**Review by damntrecky — APPROVED** [review]
Approved with a follow up on the maxDepth and comment explaining why we chose the depth.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt (hunk 2 above)
Comment (arangatang) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt:642]: Not much value imo, more importantly is that we have tests for nested and poms not in the root which we both have.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt (hunk 1 above)
Comment (arangatang) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt:649]: This is just a best guess atm, we do not have statistics for this atm.

**Review by Will-ShaoHua — APPROVED** [review]

**Pull request merged**